### ISSUE TITLE
undefined is not an object evaluating _this.props.navigation

### ISSUE THREAD

**Issue by saeedhei** [react-community/react-navigation#2198]
I'm using `DrawerNavigator` and I have 3 pages: Router page, mainScreen and a photos page,
I maked a header navbar area and I used This `<TouchableHighlight onPress={() => this.props.navigation.navigate('DrawerOpen')}>` to open Drawer menu in mainScreen and used that for  photos page too, menu is ok in mainScreen But when I click `<TouchableHighlight onPress={() => this.props.navigation.navigate('DrawerOpen')}>` in photos page, I got this error:
![photo_ - - _ - -](https://user-images.githubusercontent.com/25050762/28416078-c1b6ed18-6d67-11e7-9668-a596aab69cec.jpg)
How Can I fix That?
**my photos page:**
```
import React from 'react';
import { Button, ScrollView, View, Text, StyleSheet, TouchableHighlight } from 'react-native';
import { DrawerNavigator } from 'react-navigation';
import MaterialIcons from 'react-native-vector-icons/MaterialIcons';
import Icon from 'react-native-vector-icons/FontAwesome'

const MyNavScreen = ({ navigation }) => (
  <View>
    <View style={styles.containerNavbar}>
      <TouchableHighlight onPress={() => this.props.navigation.navigate('DrawerOpen')}>
        <Icon name="bars" size={30} color="#fff" />
      </TouchableHighlight>
      <Text style={styles.navbarTitle}>Photos</Text>
    </View>

    <ScrollView>
      <View><Text>photo</Text></View>
      <Button onPress={() => navigation.goBack(null)} title="Go back" />
    </ScrollView>
  </View>
);

const MyPhotosHomeScreen = ({ navigation }) => (
  <MyNavScreen navigation={navigation} />
);
MyPhotosHomeScreen.navigationOptions = {
  title: 'Photos',
  drawerIcon: ({ tintColor }) => (
    <MaterialIcons
      name="photo"
      size={24}
      style={{ color: tintColor }}
    />
  ),
};
export default MyPhotosHomeScreen;
```
**mainScreen:**
```
export default class MainScreen extends React.Component {
    static navigationOptions = {
        drawerLabel: 'Home',
        drawerIcon: ({ tintColor }) => (
            <MaterialIcons
                name="home"
                size={24}
                style={{ color: tintColor }}
            />
        )
    };

    render() {
        return (
            <View>
                <View style={styles.containerNavbar}>
                    <TouchableHighlight onPress={() => this.props.navigation.navigate('DrawerOpen')}>
                        <Icon name="bars" size={30} color="#fff" />
                    </TouchableHighlight>
                    <Text style={styles.navbarTitle}>mainScreen</Text>
                </View>

                <View>
                    <View style={styles.containerFooter}>
                        <Text style={styles.footerTitle}>Footer</Text>
                    </View>
                </View>
            </View>

        )
    }
}
```

**Comment by saeedhei** [comment]
Someone in stackovelflow [Said](https://stackoverflow.com/a/45227393/7428622), Change `<TouchableHighlight onPress={() => this.props.navigation.navigate('DrawerOpen')}>`
to
`<TouchableHighlight onPress={() => navigation.navigate('DrawerOpen')}>`
in `photos page` and my problem solved.

- now is my code okey without this.props?

- or I Should Access `this.props` to my `photos page` and pass `this.props.navigation.navigate('DrawerOpen')`?

**Comment by rochapablo** [comment]
+1

**Comment by spencercarli** [comment]
The difference between how you access `navigation` in the photos page and main page comes down to how you're defining the function.

With the main screen you're extending `React.Component`, thus you'll want to access any props via `this.props`.

For your profile screen you're using a stateless component, thus the props are passed via function arguments. Negating the need for `this`.
`const MyNavScreen = ({ navigation }) => (`

Does that help clarify?

**Comment by saeedhei** [comment]
Thank you

**Comment by bimix** [comment]
Cant someone just give a proper solution?

**Comment by randyqiu** [comment]
see comment on [ISSUE 832](https://github.com/react-community/react-navigation/issues/832#issuecomment-345519151)

You should pass the `navigation` prop to the screen component of navigator YOURSELF.

**Comment by Jhonsebas77** [comment]
But How? can you make a Example of that, i'am so confused,

**Comment by randyqiu** [comment]
Create a function that accept navigation as input, pass navigation to the component your want to render.

```
import PlayerView from 'player-view';
const PlayerScreen = ({ navigation }) => (<PlayerView navigation={navigation}/>);
```

**Comment by steventiro** [comment]
I have the same problem
can someone explain the solution in my code?
i try to make an icon on header which navigate to another screen
Here is my code: 
`    const HargaMenu = StackNavigator({
            hargabarang: {
                screen: HargaScreen,
                navigationOptions: {
                    title: 'Barang',
                    headerRight:
                        <Icon
                            name='search'
                            onPress={() => this.props.navigation.navigate('cariharga')}
                        />
                }
            },
            cariharga: {
                screen: CariHargaScreen,
                navigationOptions: {
                    title: 'Cari Barang',
                },
            }
        });`

can someone please help...

**Comment by lordramdhan** [comment]
const { navigate } = this.props.navigation;

 <TouchableOpacity  onPress={() => navigate("topicDetail")}>   
     <View  >
        <Text  style= {styles.Tag}> <Icon name="tags" size={14}  /> {tag.name}</Text>
     </View>
  </TouchableOpacity>

working for me

**Comment by leduytoan1988** [comment]
const {navigate} = this.props.navigation;

        return(
        <View style={styles.topbar}>
            <View style={{flexDirection: 'row',justifyContent: 'space-between',} }> 
            <TouchableOpacity onPress={() => {props.navigation.navigate('DrawerOpen')}}> 
                <Image 
                source={require('../images/menu-icon.png')} 
                style={[styles.icon]} 
                />
            </TouchableOpacity>


It said that undefined is not an object. How can I do. I create Header.js 
======
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 * @flow
 */

import React, { Component } from 'react';
import {
  Platform,
  StyleSheet,
  Text,
  View,
  Image,
  TouchableOpacity,
  Button,
  StatusBar
} from 'react-native';


//Drawer open Butoon
// const MenuButton = (props) => (
//   <TouchableOpacity onPress={() => {props.navigation.navigate('DrawerOpen')}}> 
//       <Image 
//       source={require('../images/menu-icon.png')} 
//       style={[styles.icon]} 
//       />
//   </TouchableOpacity>
// );

//End of Drawer open Butoon

export default class Header extends React.Component {

// Hidden Status Bar
    componentWillMount() {
        StatusBar.setHidden(true);
    }  

    static navigationOptions =({navigation}) =>( {
        title: 'Home',
        headerLeft: <MenuButton navigation={navigation} />,
        headerBackTitle: null,
        // Note: By default the icon is only shown on iOS. Search the showIcon option below.
        tabBarIcon: ({ tintColor }) => (
          <Image
            source={require('../images/home.png')}
          />
        ),
      });

    render(){

        const {navigate} = this.props.navigation;

        return(
        <View style={styles.topbar}>
            <View style={{flexDirection: 'row',justifyContent: 'space-between',} }> 
            <TouchableOpacity onPress={() => {props.navigation.navigate('DrawerOpen')}}> 
                <Image 
                source={require('../images/menu-icon.png')} 
                style={[styles.icon]} 
                />
            </TouchableOpacity>
             <Text > Wearing a dress </Text>
            <Image source={require('../images/dress-icon.png')} style={[styles.icon]}  />
            </View>
        </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
      flex: 1,
      flexDirection: 'row',
    },
    topbar: {
      height: '10%',
      backgroundColor: '#34B089',
    },
    icon: {
        width: 56,
        height: 56,
      },
  });
  ======
and I call it in Home.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 * @flow
 */

import React, { Component } from 'react';
import {
  Platform,
  StyleSheet,
  Text,
  View,
  Image,
  TouchableOpacity,
  Button
} from 'react-native';

import Header from '../Main/Header.js'

// Drawer open Butoon
// const MenuButton = (props) => (
//   <TouchableOpacity onPress={() => {props.navigation.navigate('DrawerOpen')}}> 
//       <Image 
//       source={require('../images/menu-icon.png')} 
//       style={[styles.icon]} 
//       />
//   </TouchableOpacity>
// );

//End of Drawer open Butoon

export default class Home extends React.Component<{}> {
 
 
  static navigationOptions =({navigation}) =>( {
    title: 'Home',
    // headerLeft: <MenuButton navigation={navigation} />,
    headerBackTitle: null,
    tabBarLabel: 'Home',
    // Note: By default the icon is only shown on iOS. Search the showIcon option below.
    tabBarIcon: ({ tintColor }) => (
      <Image
        source={require('../images/home.png')}
        style={[styles.icon]}
      />
    ),
  });

  render() {

    // const {navigate} = this.props.navigation;

    return (
      <View style={styles.container}>
        <Header />
        <View style={styles.body}>
          <Button 
            title="Touch me"
            onPress={() => {
              navigate('Search', {username:'From Details'});
            }}
          />
        </View>
        </View>
    // <View style={styles.container}>

    //   <View style = {styles.top}>
    //     <View style={styles.profileimage}> 
    //     </View>
    //   </View>



    // </View>
    
    
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  
  topbar: {
    height: '10%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'red',
  },
  
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
  icon: {
    width: 26,
    height: 26,
  },
  body:{
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
  top: {
    height: '45%',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#98d2c1',
  },

  profileimage:{
    width: 140,
    height: 140,
    borderRadius: 100,
    borderColor: '#fff',
    backgroundColor: "#eee",
  }
});

**Comment by muganwas** [comment]
I had the same issue when I passed props to an element.
I had a constructor but hadn't passed props to it. ie `
 constructor(){
super()
this.state={
data: this.props.info
}
}` When I passed props to the constructor, it was resolved `
constructor(props){
super(props)
}`

**Comment by wyz1** [comment]
This is how i solved this issue. When using a class extending the Component class, you have to declare the navigation within the constructor like this:

`   constructor(props){
      super(props)
      this.state = {
        navigation: this.props.navigation //here you assign the navigation props to a state
      }
    }`

You can now use: this.state.navigation object within your code, like this 

`<TouchableOpacity
          onPress={ ()=> this.state.navigation.navigate('Home', {name: 'Victor Osas Ighalo'}) } 
          >`
I hope this helps someone.

**Comment by gabrielsla3** [comment]
Hello,

I solved this passing navigation props to my components and using it there. For example:

I have this in my StackNavigator:

```
export const NavigationApp = StackNavigator({
  Home: { screen: CenaPrincipal },
  CenaClientes: { screen: CenaClientes },
  CenaContatos: { screen: CenaContatos }
}, {
    headerMode: 'none',
  });
```

In my 'Home' class, I've `const { navigate } = this.props.navigation;`

And when I navigate between the pages, I passed the props through params like this:

```
<TouchableOpacity onPress={() => navigate('CenaClientes', { navigation: this.props.navigation })}>
                            <Image style={styles.imgMenu} source={menuCliente} />
                        </TouchableOpacity>
```

And in CenaClientes, I used the navigation props, so can go back between the pages or access other pages normally:

`const { navigate } = this.props.navigation;`

So, I guess that not is the better way to do this. No make sense to me all this work to do the simple route between the pages, but, this way works to me.

I hope this helps.

**Comment by vksgautam1** [comment]
this is what i tried and it worked like charm
[code]
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import Home from './src/home';
import Screen1 from './src/Screen1';
import {  StackNavigator} from 'react-navigation';
import {  AppRegistry } from 'react-native';

export const Stack = new StackNavigator({
  Home: { screen: Home },
  Screen1:{ screen: Screen1 }
},{
  initialRouteName:'Home',
}
    
)

export default class App extends React.Component {
 
  render() {
     return  <Stack />;   
           
  }
} 
[/code]

and in other screen
[code]
render() {
    const { navigate } = this.props.navigation;
    return (
      <View style={styles.container}>
        <Text>Open upapp cool !</Text>   
        <Text>Changes you makefgdg will automatically reload.</Text>
        <Text>Shake your phone to openhvffff the developer menu.</Text>
        <View>
        <Button  title="click" 
          onPress={() => navigate('Screen1')}
          />
</View>   

      </View>  
    );
  }
[/code]

**Comment by arnoldbird** [comment]
What would @randyqiu solution [above](https://github.com/react-navigation/react-navigation/issues/2198#issuecomment-348045057) work in an ES6/7 class rather than a functional component?

**Comment by berryman17** [comment]
This took me a while and my problem was not exporting the correct Component. I found that I must export my App component which contains the RootStack (if you went through the tutorial). I had previously been exporting another component due to my web application habits.

`
export default class App extends React.Component {
  render() {
    return <RootStack />;
  }
}
`

**Comment by brentvatne** [comment]
hey folks, if you are still running into any issues here please create a new issue that follows the issue template. thank you! sorry for locking but i just want to be sure that your concerns are heard and it's too hard to follow a big thread.
